Test grade

meitals91 opened the pull request:
Create test for the model class - Grades list & GPA calculator and made migration files for creating test data for your DEV environment called "0002_test_data.py"

fix #26 
fix #72

meitals91 committed:
Adding def str to course class

simzacks commented on grades/models.py:
```diff
@@ -24,7 +24,7 @@ class Grade(models.Model):
     @classmethod
     def addGradeUserCourse(cls, user, courseID, grade, semester, year):
-        newGrade = Grade.object.create()
+        newGrade = Grade.objects.create()
```
While this will work, it is cleaner to use cls.objects.create() instead of the class name itself.
Two reasons for this:
1) If you change the name of the class you will have to change it here as well.
2) If you have a child class and you want this to work on it as well, it should use the object passed in.

While those reasons may not apply in this specific case, it is still goo practice.

simzacks replied:
One other point, at the end of the function, you probably need to save.

simzacks replied:
Another thought is you might want to use newgrade = Grade() and not Grade.objects.create(). If this works (I'm not sure it will), it will actually create a record with no fields other than the primary key, then when you add the additional fields and call save(), that will then do an update, so it's 2 DB calls instead of 1.

simzacks replied:
In the Grade model you are allowing both User and Course to be null and blank. Can you think of a situation where you would want to have a grade record but not know who the user is or what course it is?

meitals91 replied:
@simzacks I have changed the models as you mentioned above.
About your last point I did it as it has shown in your presentation about the AUTH_USER_MODEL field.  I fixed it now.

simzacks commented on grades/models.py:
```diff
@@ -34,5 +34,5 @@ def addGradeUserCourse(cls, user, courseID, grade, semester, year):
     @classmethod
     def removeCourse(cls, courseID):
-        Course.object.filter(id=courseID).delete()
+        Course.objects.filter(id=courseID).delete()
```
Shouldn't this be a classmethod of the Course object?
Also, see comment above on using cls instead of Course.

simzacks commented on grades/tests.py:
```diff
@@ -1,3 +1,16 @@
+    def test_addGradeUserCourse(cls):
+        out = Grade.objects.all()
+        assert isinstance(out, QuerySet)
+        assert all(isinstance(g, Grade) for g in out)
```
The previous 2 asserts are testing the base django functionality. No real need to do this.

simzacks commented on grades/migrations/0002_test_data.py:
```diff
@@ -0,0 +1,31 @@
+            (user1, course1, 85, 'winter', 2021),
+            (user1, course2, 60, 'spring', 2020),
+            (user1, course1, 20, 'summer', 2090),
+        ]
```
Did you verify this works? I would think that you would have to pass in A, B or C instead of winter, spring, summer

simzacks commented on grades/tests.py:
```diff
@@ -1,3 +1,16 @@
+            (1, 1, 85, 'winter', 2021),
+            (1, 2, 60, 'spring', 2020),
+            (1, 1, 20, 'summer', 2090),
+        ]
```
As mentioned in test data, the values it should return for semester are A, B and C.

This will work as an initial test based on the assumption that the migration data is the only data in the system. However, a test module shouldn't assume that specific data is in the system unless that is part of the production system that it is testing. For the test module to be stand-alone, the data that you are verifying should be added in the test module itself, and should use a filter so you get the specific data you are looking for instead of using .all

meitals91 replied:
@simzacks You are right, that was my intention. I saw now that the choises didn't work as i wanted them to work. I saw in your presentation what you did in 'Gender' class and made changes at my class according to that.

meitals91 commented:
@simzacks  @ravidbro @Omeramsc  I updated the test model as you requested. I olso fixed issueon the migration becuse in the last time was created a file xxxx_initial_ so i deleted it.

simzacks commented on grades/models.py:
```diff
@@ -9,30 +9,44 @@ class Course (models.Model):
     year = models.IntegerField(default=0)
 
+    @classmethod
+    def addCourse(cls, course_name, credinitials):
```
This should be a classmethod of course, not of grade.

meitals91 replied:
Did you meant that it supposed to be under the class Course? @simzacks

simzacks replied:
yes. The way you're going to call this is classname.classmethod, so the way you have it set up now, it would be
`Grade.addCourse(...) `
instead of
`Course.addCourse(...)`

simzacks commented on grades/models.py:
```diff
@@ -9,30 +9,44 @@ class Course (models.Model):
+        return newCourse
+
+    @classmethod
+    def removeCourse(cls, courseID):
```
This should be a classmethod of course, not of grade.

meitals91 replied:
Did you meant that it supposed to be under the class Course? @simzacks

simzacks replied:
same as above

simzacks commented on grades/tests.py:
```diff
@@ -1,3 +1,50 @@
+        c = Course(course_name="java", credinitials="3")
+        c.save()
+        c = Course.objects.get(id=3)
+        c.delete()
```
You created a classmethod for delete course. Why not use that?

meitals91 replied:
@simzacks Yor are right , I changed it now

simzacks replied:
I don't see the change

meitals91 replied:
I didnt commit the changes cause i was waiting for your commets. I do it now @simzacks

simzacks commented on grades/tests.py:
```diff
@@ -1,3 +1,50 @@
+        c = Course(course_name="NLP", credinitials="3")
+        c.save()
+        g = Grade(user=user3, course=c, grade=90, semester='A', year=2021)
+        g.save()
```
You have a classmethod called addGradeUserCourse which does the same thing as this. Why not use that?

simzacks commented on grades/tests.py:
```diff
@@ -1,3 +1,50 @@
+    def test_removeGrade(cls):
+        user3 = User.objects.create_user(username="idan", password="123456")
+        c = Course(course_name="NLP", credinitials="3")
+        c.save()
```
You defined a classmethod called addCourse, why not use that?

meitals91 replied:
@simzacks Use it now it , I update now the commit

simzacks commented on grades/tests.py:
```diff
@@ -1,3 +1,50 @@
+class TestCourse:
+    def test_addCourse(cls):
+        c = Course(course_name="java", credinitials="3")
+        c.save()
```
same comment as above

meitals91 replied:
@simzacks Use it now it , I update now the commit

simzacks commented on grades/tests.py:
```diff
@@ -1,3 +1,50 @@
+        c.save()
+        g = Grade(user=user3, course=c, grade=90, semester='A', year=2021)
+        g.save()
+        g = Grade.objects.get(id=2)
+        g.delete()
```
You have a classmethod defined, Grade.removegrade(gradeID). Why not use that?

meitals91 replied:
@simzacks Use it now it , I update now the commit

simzacks reviewed (changes requested):
You don't need the classmethods: addCourse, removeCourse, addGradeUserCourse, removeGrade. But if you define them, you should use them.

meitals91 commented:
@simzacks I updated the commit

meitals91 commented:
> Consider retrieving existing files

@Yarboa Can you please explain what you mean i'm not sure i understand you.

simzacks reviewed (approved):
Looks good to me.

lmilbaum reviewed (changes requested):
Too many commits in this PR. Please squash them to one commit.

meitals91 commented:
@lioramilbaum  If i want to squash the commits as you say. I choose 's' only the commits above? because there are other commits that the other team member did and I dont want to squash them?

meitals91 commented:
@lioramilbaum  I have squashed the last 4 commits to one. is ot ok?

lioramilbaum commented:
> @lioramilbaum I have squashed the last 4 commits to one. is it ok?

This issue was discussed over the session today. Maybe the easiest approach is to close this PR and start over with a new branch based on master.

meitals91 commented:
> > @lioramilbaum I have squashed the last 4 commits to one. is it ok?
> 
> This issue was discussed over the session today. Maybe the easiest approach is to close this PR and start over with a new branch based on master.

@lioramilbaum Ok, I will fix it

simzacks commented on grades/models.py:
```diff
@@ -16,27 +16,45 @@ class Course (models.Model):
 
+    @classmethod
+    def addCourse(cls, course_name, credinitials):
+        newCourse = Course()
```
Your course model now has a user attached to it as well.
This is a change @or313 made in PR #67

simzacks commented on grades/migrations/0002_test_data.py:
```diff
@@ -0,0 +1,29 @@
+
+        user1 = User.objects.create_user(username="ido", password="123456")
+        course1 = Course.objects.create(course_name="course1", credinitials="3")
+        course2 = Course.objects.create(course_name="course2", credinitials="2")
```
Course now has a user field, as well

ravidbro commented:
you have conflicts on 'initial' file, it means that your file is not initial any more, it's better to delete your changes from this file in this PR and run 'make migrations' again to generate a new file.
the migration files are supposed to be added between PRs and not changed

meitals91 commented:
@simzacks  I have update the changes in my model and test.

meitals91 commented:
@ravidbro I fixed the confilct, I did a new  migrate again.

meitals91 commented:
@lioramilbaum I have squashed the commit to one as yoo requested
